rcl/include/rcl/client.h

```diff
@@ -0,0 +1,300 @@
+
+#ifndef RCL__CLIENT_H_
+#define RCL__CLIENT_H_
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include "rosidl_generator_c/service_type_support.h"
+
+#include "rcl/macros.h"
+#include "rcl/node.h"
+#include "rcl/visibility_control.h"
+
+/// Internal rcl client implementation struct.
+struct rcl_client_impl_t;
+
+/// Handle for a rcl client.
+typedef struct rcl_client_t
+{
+  struct rcl_client_impl_t * impl;
+} rcl_client_t;
+
+/// Options available for a rcl client.
+typedef struct rcl_client_options_t
+{
+  /// Middleware quality of service settings for the client.
+  rmw_qos_profile_t qos;
+  /// Custom allocator for the client, used for incidental allocations.
+  /* For default behavior (malloc/free), use: rcl_get_default_allocator() */
+  rcl_allocator_t allocator;
+} rcl_client_options_t;
+
+/// Return a rcl_client_t struct with members set to NULL.
+/* Should be called to get a null rcl_client_t before passing to
+ * rcl_initalize_client().
+ * It's also possible to use calloc() instead of this if the rcl_client is
+ * being allocated on the heap.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_client_t
+rcl_get_zero_initialized_client(void);
+
+
+/// Initialize a rcl client.
+/* After calling this function on a rcl_client_t, it can be used to send requests of the given
+ * type by calling rcl_send_request().
+ * Once a response has been sent by the service, the response is available to the client when
+ * rcl_take_response() is called.
+ *
+ * The given rcl_node_t must be valid and the resulting rcl_client_t is only
+ * valid as long as the given rcl_node_t remains valid.
+ *
+ * The rosidl_service_type_support_t is obtained on a per .srv type basis.
+ * When the user defines a ROS service, code is generated which provides the
+ * required rosidl_service_type_support_t object.
+ * This object can be obtained using a language appropriate mechanism.
+ * \TODO(wjwwood) probably should talk about this once and link to it instead
+ * \TODO(jacquelinekay) reworded this for services with substitutions, should it refer to messages?
+ * For C this macro can be used (using std_msgs/String as an example):
+ *
+ *    #include <rosidl_generator_c/service_type_support.h>
+ *    #include <example_interfaces/srv/add_two_ints.h>
+ *    rosidl_service_type_support_t * ts =
+ *      ROSIDL_GET_SERVICE_TYPE_SUPPORT(example_interfaces, AddTwoInts);
+ *
+ * For C++ a template function is used:
+ *
+ *    #include <rosidl_generator_cpp/service_type_support.hpp>
+ *    #include <example_interfaces/srv/add_two_ints.h>
+ *    rosidl_service_type_support_t * ts = rosidl_generator_cpp::get_service_type_support_handle<
+ *      example_interfaces::srv::AddTwoInts>();
+ *
+ * The rosidl_service_type_support_t object contains service type specific
+ * information used to send or take requests and responses.
+ *
+ * \TODO(wjwwood) update this once we've come up with an official scheme.
+ * The topic name must be a non-empty string which follows the topic naming
+ * format.
```

Comment:
I think the term "topic" is specific to pub/sub, and we should use "service" name instead. Services may also have a different naming format since topics are content oriented and services could also be location oriented, i.e. `service://node/service_name` versus `service://*/service_name` (or something like that).